### ISSUE TITLE
Support floating point differences between .NET Framework and Core

### ISSUE THREAD

**Issue by heaths** [Azure/azure-sdk-for-net#27695]
Tangentially related to #27688 we might consider some solution to use separate recordings for .NET Framework and .NET (Core) in cases e.g., floats are serialized differently. Pavel and I had looked into other solutions including what a .NET team themselves was doing, but the only viable solution (without refactoring a bunch of production code) wasn't exactly rock solid (JSON decoder that heuristically assumes a certain rounding and precision of floats). @m-nash had an idea to use separate files, though.

As one idea, we could add a property to `RecordedTestBase` to support this somehow. Most simplistically is just a bool to pivot on .NET Framework vs. .NET but perhaps we can go a little further and support other pivots. What we would ideally have to figure out is how to handle which platforms need to record tests. Currently, I believe this is done manually by the SDK dev. For example, when I re-record tests I only do so for the latest target e.g., net6.0. Using the pivot, could we just let SDK devs run all (in record mode) and figure out for which platforms we actually need to run / record?

/cc @JoshLove-msft @christothes

**Comment by JoshLove-msft** [comment]
That said, I would like to avoid adding this feature if we can come up with another way to fix the floating point issue.

**Comment by heaths** [comment]
When Pavel and I were looking into this, we scrounged around several repos in the dotnet org and found a few issues, and I found one solution I didn't bookmark and couldn't find again but we could try looking again. IIRC, it was an aspnet repo. 99% sure it was under the dotnet org account still. That was the basic solution mentioned above about the JSON decoder. You could set that in tests without having to refactor our whole JSON deserialization infrastructure...which is still an option, but perhaps not a great one considering the risk to production code. Ideally, we were hoping to avoid that and just "plug in" a solution for tests only.

**Comment by christothes** [comment]
could this be implemented with something similar to a sanitizer?

**Comment by JoshLove-msft** [comment]
> could this be implemented with something similar to a sanitizer?

Yes, you could do something like conditional sanitization depending on the platform.

**Comment by heaths** [comment]
I suppose now that we have the `JsonPathSanitizer` that would be an easy way to go. Pavel and I had talked about sanitizing the path but at the time you still had to deserialize the content yourself in your own sanitizer. It felt more heavy-handed compared to just ignore bodies as needed.

**Comment by heaths** [comment]
From offline discussion, we could just conditionally disable body comparisons.

**Comment by m-nash** [comment]
This recently came up again with an internal customer who wanted to parallelize the tests which means they needed a different "name" per framework and they implemented that [this way](https://github.com/Azure/azure-sdk-for-net/blob/b280663bbf3ac899e215239eb99e9d06505083c5/sdk/chaos/Azure.ResourceManager.Chaos/tests/ChaosManagementTestBase.cs#L71-L86).  This works great for recording and live tests and speeds things up since most of the test time is waiting for the service to do things.

The issue came in playback mode and they added [this workaround](https://github.com/Azure/azure-sdk-for-net/pull/33283/files#diff-66b31aca86d05d787944c6166020a76d752e38dd6bbb35a599342c5a18fee813R73-R76) for now, but separate recordings would solve this.

Are there any compatibility issues we would be missing if we isolated record and playback to each dotnet target?  Is there any advantages to playing back the same recording across all dotnet targets?

**Comment by m-nash** [comment]
> > Are there any compatibility issues we would be missing if we isolated record and playback to each dotnet target? Is there any advantages to playing back the same recording across all dotnet targets?
> 
> The main advantage is that we avoid multiplying the number of recordings that we store by the number of targets.

This makes sense and should be mitigated with the change to store the recordings outside of the repository.

**Comment by JoshLove-msft** [comment]
Mitigated sure, but I don't think it means we should record all targets by default. We could have an attribute that can be applied as needed.

**Comment by heaths** [comment]
I agree we shouldn't necessarily optimize for recording parallel test runs. First, we don't necessarily want to multiply the number of test recordings - *especially* when still in the code repo. Second, you end up with a mix of different headers and potentially behaviors (hopefully not) in recording files. By default, we should record on the latest (now net6.0, but soon net7.0) in the test proxy when we do that, and probably add to our _CONTRIBUTING.md_ for people to only record on latest. That's what I always do. Too many synchronization issues otherwise.

With an overlap in files, once the files are in place there should be no good reason you can't run tests in parallel. I generally do and have never had a problem (as long as my test infrastructure code doesn't have any sync issues itself).

**Comment by m-nash** [comment]
@heaths that is an interesting suggestion to ONLY record on latest.  That would solve the customers issue above for recording, but still not in live tests.  I think this is where the real value of parallelization comes in to reduce the time it takes to run our nightly runs.  We have run into countless pipeline timeout issues when trying to add mgmt plane tests into the nightly runs because everything is run serially.

**Comment by heaths** [comment]
Being able to run tests - especially if you're bifurcating on data- vs. management-plane seems like a separate issue with separate problems, though. For example, maybe we need separate pipelines for each so that management-plane tests can have a longer run time without adversely affecting data-plane tests. But live tests don't record, so this seems like a wholly separate concern we could track separately. If you're _recording_ in parallel yeah, it's related if we did want to record separately for each test TFM, but I don't think we should. (Perhaps we should change the title of this issue in that regard, too.)

**Comment by heaths** [comment]
Changing title based on discussion. Recording for separate frameworks is one way to do that, but there's some pretty significant downsides to that. We should consider all possibilities.

One other suggestion we considered offline, IIRC, was to make the differ a little smarter and either throw an exception we could heuristically ignore, but we hadn't figured out a way yet. For one, we'd have to stop comparing response Content-Length which makes every comparison a little longer. Maybe we could opt into that, though.

**Comment by m-nash** [comment]
Separate pipelines doesn't solve the time it takes to run live tests, it just removes a noisy neighbor issue where mgmt plane is adversely impacting data plane.

**Comment by heaths** [comment]
True, but from what I've been told, much of that time spent doing management plane testing can't really be helped.

FWIW, tests themselves can be parallelized if possible. NUnit defaults to non-parallelized, but you can put `[Parallelized]` on some base class. I do this for many of mine, with occasional `[NonParallelized]` on some that can't. As long as your resource names are generated using the `Recording.Random` it should work fine.

Running TFM in parallel is a separate matter, though. Right now it's not because that's safest. I've talked with @weshaggard and @benbp about this in the past but we didn't get too far in making it possible. Maybe it's worth talking about further, but we may still want a separate issue tracking it. I'd like to keep this issue focused on the body comparison issue as it most pertains to - but not limited to - floating points.

**Comment by m-nash** [comment]
`Recording` class is not thread safe so that doesn't work.  If this is off topic for this issue lets move it somewhere else didn't mean to derail this as the original title made the issues seem similar.

**Comment by heaths** [comment]
That said, this wasn't even about parallelism. Floating point de/serialization differs across .NET Framework and .NET nee, "Core". Recording separate files was one way, but I probably shouldn't have made the title suggest it was the only way.

**Comment by pallavit** [comment]
@m-nash , @JoshLove-msft , @heaths , @christothes given we have now moved the recordings outside the repo, do you see us have a clear path forward here i.e. have separate recordings for different TFMs? Given we have a workaround available - should we implement it as a first class support or simply include the sample for anyone who wants to use this (we expect the usage here to be minimal).

**Comment by heaths** [comment]
Separate recordings per TFM is something I've been wanting to drive after some discussions late last year with @JoshLove-msft et. al. - basically, we record on all supported/desired TFMs and playback during CIs, but do live tests on the latest only since we're testing the SDKs not the services. Services that want to use our tests for prod could opt to run all versions in lieu of any other tests they may have otherwise written. I think that jives better with the shift left aspects remaining of our ongoing API first strategy, yes? /cc @jsquire

**Comment by jsquire** [comment]
> Do you agree? /cc @jsquire

I do.  This seems like a good trade off of some storage space for quicker test runs and non-trivial cost savings.

**Comment by heaths** [comment]
Well, CIs may actually take a little longer. Depends how much they are parallelized. But this would definitely and significantly reduce live testing times, which also reduces costs. Just want to set accurately set expectations. :)

**Comment by jsquire** [comment]
Yeah, I was thinking of live runs.
